### DISCUSSION TITLE
[FEATURE] Validação das Informações conforme Layout do SPED

### DISCUSSION THREAD

**Discussion by orochasamuel** [orochasamuel/fiscalbr-net#86] [Ideas]
Galera,

Recentemente tivemos uma issue sobre validação das informações na lib (vide issue #85).

O que vocês acham da ideia de incluir a validação conforme o layout exemplifica?

Exemplo:

C500 na EFD Fiscal

![image](https://user-images.githubusercontent.com/15462690/224048989-653af67c-6c8d-4187-972e-d015f334714e.png)

Para operações de ENTRADA o Indicador Destinatário / Acessante não deve ser informado. Porém, para operações de SAÍDA é obrigatório, **ou seja, como pode ser usado nas duas operações se o valor não for informado a lib não vai acusar a falta deste campo mesmo que a operação seja saída (e nesse caso é obrigatório).**

Atualmente não temos esse tipo de validação na lib, tem validação estrutural para informar caso algum campo esteja mal formatado.

Para validar esse tipo de situação conforme o layout exemplifica teríamos que ter um método de validação classe a classe (vai tempo para incluir isso).

Alguém já fez algo do tipo? Vocês fazem validação a moda antiga (if isso else aquilo)? Alguém pode sugerir algo diferente?

@marcosgerene @marcianobandeira @adrianotrentim @ebitencourt @ThalyaAutocom @PGama-Rodrigo @hbdbim @EderWillian @eduarda80

**Comment by eduprog** [comment]
Eu penso ser muito utilt, pois se para gerar dentro dos parâmetros está no manual, creio que para este o componente pode tranquilamente fazer.esta validação, de forma a ter um retorno com os erros, blocos e campos que ocorreram, que poderatser validado pelo dev, assim naotcausa impacto para quem está montando o arquivo.

Ou seja, iremos trabalhar com domain notifications, ficaria top..

Vou pedir aos devs da empresa para passar como usamos.

**Comment by marcianobandeira** [comment]
Boa tarde senhores, eu penso que algumas coisas mais simples são positivas de se fazer esse trabalho. Porém, a legislação que envolve os speds são muito interpretativas, há normas diferentes em cada estado, há situações em que o contribuinte tem alguma particularidade devido a um processo judicial e coisas do tipo.
No meu ver, seria algo complexo de se conseguir abarcar todos os pontos, até mesmo porque a legislação está em constante mudança.

**Comment by WashingtonAutocom** [comment]
@marcosgerene penso ser útil também, de qualquer forma ficaria como é determinado no layout não temos outro caminho, ficaria mais claro na hora de preencher.

Como comentado pelo @marcianobandeira. 
Pode ser trabalhoso mesmo, mais se houver qualquer mudança na legislação que impacte como as informações que devem ser apresentadas no arquivo do SPED, tem que ser gerado uma nova versão do layout, ai é natural que implementamos a alteração para adaptar a nova versão.

**Comment by marcosgerene** [comment]
Acredito que TODA E QUALQUER validação deva ser opcional, até mesmo as obvias (campo X é obrigatório). Tenho casos onde preciso gerar um arquivo, importar em outra aplicação, adicionar informações e exportar.

Um exemplo de validação que só me atrapalha: O usuário não configurou os dados do contador. O componente me retorna erro porque o nome dele está vazio.

Ou seja, hoje eu iria no caminho inverso, removeria o que já existe ao invés de adicionar mais informações. Na minha visão de uso de componentes tudo que for regra de negócio é problema meu, não gosto quando o componente "se mete onde não é chamado" hahaha

Agora sobre a sugestão, acho válido, mas acredito que deva ser opcional. E neste caso eu desligaria TUDO quando a opção fosse por não validar.

**Reply by orochasamuel** [reply]
@marcosgerene o método "EscreverCampos" tem várias assinaturas diferentes, em uma delas é possível usar o parâmetro para ignorar os erros e gerar normalmente a linha.

```cs

            // aqui dispara exception
            //var currentResult = Common.Sped.EscreverCamposSped.EscreverCampos(source, CodigoVersaoLeiaute.V17);

            // aqui não dispara nenhuma exception
            var currentResult2 = Common.Sped.EscreverCamposSped.EscreverCampos(
                source,
                CodigoVersaoLeiaute.V17,
                DateTime.Now,                       // data de competência do arquivo
                false,                              // remove espaços em branco no começo e no final da string
                true                                // ignora os erros apresentados na escrita
                );

```

**Reply by orochasamuel** [reply]
Escrevi um teste para validar a escrita sem os dados e ignorando os erros.

```cs

public void Escrever_Registro_0100_Sem_Dados_Contador_EFDFiscal()
        {
            var expectedResult =
                $"|0100||||||||||||||{Environment.NewLine}";

            var source = new EFDFiscal.Bloco0.Registro0100
            {

                Nome = "",
                Cpf = "",
                Crc = "",
                Cnpj = "",
                Cep = "",
                End = "",
                Num = "",
                Compl = "",
                Bairro = "",
                Fone = "",
                Fax = "",
                Email = "",
                CodMun = "",
                
            };

            var currentResult = Common.Sped.EscreverCamposSped.EscreverCampos(
                source,
                CodigoVersaoLeiaute.V17,
                DateTime.Now,
                false,
                true
                );

            currentResult = currentResult.Replace(Constantes.StructuralError, "");

            Assert.Equal(expectedResult, currentResult);
        }

```

**Reply by marcosgerene** [reply]
@orochasamuel 

Eu não chamo o ```EscreverCampos()``` diretamente, utilizo da seguinte forma:
```
var sped = new ArquivoEFDFiscal();
/*
  preenchimento dos objetos
*/
sped.GerarLinhas();
sped.CalcularBloco9();
sped.Escrever(path);
```

Na época que fiz o GerarLinhas acredito que não me atentei ou até mesmo não existia essa opção, vou alterar para ele ter a opção de validar ou não, no meu caso eu não quero validação alguma, no caso dos amigos acima acredito que a validação é válida.

---

Edit: Eu já faço esse tratamento! Eu não lembrava que já havia resolvido a situação, mas de qualquer forma a opinião que passei antes segue na mesma linha, toda validação, por menor que seja deve ser opcional.

Método ```GerarLinhas()``` atual:

![image](https://user-images.githubusercontent.com/15945003/224488233-3109d7da-c41d-4df9-8532-0a3a1f9a0245.png)

Neste caso eu jogo todos os erros para a variável ```List<string> Erros```. O usuário pode externalizar esses erros, tratar de alguma forma ou simplesmente ignorar (meu caso).
